documents/read_acls.go

```diff
@@ -0,0 +1,336 @@
+// if collaborators are empty, it is a no-op
+func (cd *CoreDocument) initReadRules(collaborators []identity.CentID) {
+	if len(cd.Document.Roles) > 0 && len(cd.Document.ReadRules) > 0 {
+		return
+	}
+
+	if len(collaborators) < 1 {
+		return
+	}
+
+	cd.addCollaboratorsToReadSignRules(collaborators)
+}
+
+// findRole calls OnRole for every role that matches the actions passed in
+func findRole(cd coredocumentpb.CoreDocument, onRole func(rridx, ridx int, role *coredocumentpb.Role) bool, actions ...coredocumentpb.Action) bool {
+	am := make(map[int32]struct{})
+	for _, a := range actions {
+		am[int32(a)] = struct{}{}
+	}
+
+	for i, rule := range cd.ReadRules {
+		if _, ok := am[int32(rule.Action)]; !ok {
+			continue
+		}
+
+		for j, rk := range rule.Roles {
+			role, err := getRole(rk, cd.Roles)
+			if err != nil {
+				// seems like roles and rules are not in sync
+				// skip to next one
+				continue
+			}
+
+			if onRole(i, j, role) {
+				return true
+			}
+
+		}
+	}
+
+	return false
+}
+
+// GetExternalCollaborators returns collaborators of a Document without the own centID.
+func (cd *CoreDocument) GetExternalCollaborators(self identity.CentID) ([][]byte, error) {
+	var cs [][]byte
+	for _, c := range cd.Document.Collaborators {
+		c := c
+		id, err := identity.ToCentID(c)
+		if err != nil {
+			return nil, errors.New("failed to convert to CentID: %v", err)
+		}
+		if !self.Equal(id) {
+			cs = append(cs, c)
+		}
+	}
+
+	return cs, nil
+}
+
+// NFTOwnerCanRead checks if the nft owner/account can read the Document
+func (cd *CoreDocument) NFTOwnerCanRead(tokenRegistry TokenRegistry, registry common.Address, tokenID []byte, account identity.CentID) error {
+	// check if the account can read the doc
+	if cd.AccountCanRead(account) {
+		return nil
+	}
+
+	// check if the nft is present in read rules
+	found := findRole(cd.Document, func(_, _ int, role *coredocumentpb.Role) bool {
+		_, found := isNFTInRole(role, registry, tokenID)
+		return found
+	}, coredocumentpb.Action_ACTION_READ)
+
+	if !found {
+		return errors.New("nft not found in the Document")
+	}
+
+	// get the owner of the NFT
+	owner, err := tokenRegistry.OwnerOf(registry, tokenID)
+	if err != nil {
+		return errors.New("failed to get NFT owner: %v", err)
+	}
+
+	// TODO(ved): this will always fail until we roll out identity v2 with CentID type as common.Address
+	if !bytes.Equal(owner.Bytes(), account[:]) {
+		return errors.New("account (%v) not owner of the NFT", account.String())
+	}
+
+	return nil
+}
+
+// AccountCanRead validate if the core Document can be read by the account .
+// Returns an error if not.
+func (cd *CoreDocument) AccountCanRead(account identity.CentID) bool {
+	// loop though read rules, check all the rules
+	return findRole(cd.Document, func(_, _ int, role *coredocumentpb.Role) bool {
+		_, found := isAccountInRole(role, account)
+		return found
+	}, coredocumentpb.Action_ACTION_READ, coredocumentpb.Action_ACTION_READ_SIGN)
```

Comment:
same as above, ACLReadSign

Reply:
Dont need to create already existing constants